feat(comments)

fmSand opened the pull request:
### Key Changes:

**Comment Service (services/CommentService.js):**

- Implemented methods for create, getProjectComments, update, softDelete, getOneWithDetails and getOneId.
- getOneWithDetails(commentId) is the single source for fetching a comment with full associations (user, replies with user) and returning a clean JSON object.
- softDelete now sets isDeleted: true and userId: null to anonymize the comment upon direct deletion.

**Controllers (controllers/commentController.js):**

- Uses successResponse/errorResponse helpers for consistent API response structure.
- Calculates pagination meta for the GET endpoint, including the hasNext flag.

**Routing (routes/comments.js, routes/projectComments.js):**

Split Routes: Comment routes were split for clarity and REST principles:
- projectComments.js (mounted under /projects using mergeParams: true): 
- comments.js (mounted under /comments): Handles actions on specific comments identified by their own ID (PUT /comments/:id, DELETE /comments/:id).
- Applied necessary middleware: authenticate, validateSchema, validateParamSchema, ownsEntity, asyncHandler.

### Other

- Yup schemas for creating/updating comments and validating URL parameters (projectId, commentId).
- Renamed vote route file from vote.js to votes.js.
- Routes added to Postman
- afterDestroy Hook add and fix will be in it's own PR.

fmSand committed:
feat: Implement comments functionality and API endpoints

- Add Comment model with threaded reply support and soft deletion
- Implement CommentService with CRUD operations for comments
- Create comment controllers for handling requests
- Add validation schemas for comments (create/update) and params (projectId/commentId)
- Implement routes for comments management:
  - Project-specific comment endpoints (/projects/:projectId/comments)
  - Direct comment manipulation endpoints (/comments/:id)
- Update routing configuration in app.js and routes/index.js
- Rename route files (vote → votes, comment → comments)
- Rename projectTag Model to ProjectType
- Add Postman request collection for comments API
- Optimize database queries with indexes on projectId and parentId fields

fmSand committed:
fix(comments): Nullify user info for deleted comments

fmSand committed:
refactor: Improve comment deletion system

- Refactor comment serialization into a dedicated commentSerializer module
- Clear separation between data access and response formatting
- Improve handling of deleted comments in API responses
- Fix User model hook to properly handle comment anonymization on user deletion
- Fix typo in User model indexes definition

fmSand committed:
fix: merge conflict

fmSand committed:
fix merge conflict

fmSand committed:
Update commentSerializer.js

Anclagen commented on utilities/commentSerializer.js:
I'm a bit unclear on the purpose of this function, the raw output already seems fully usable by the frontend.
- If User is null, that already signals a deleted user.
- If a comment is deleted, isDeleted is set to true, and the message is replace by default.
- Edited comments can be detected by comparing createdAt and updatedAt.
By keeping the response format consistent and simple, we give the frontend full flexibility to handle display logic without needing backend-specific serialization.

fmSand replied:
Originally was to simplify things for the frontend giving them an exact format for both comments and replies. But I  see your points!  Removed

Anclagen commented on schema/params.js:
This might need to be standardised so we have a int id schema and uuid schema but this is an issue to be address on clean up and refactoring.

fmSand replied:
Left it as is for now then, thank you.

Anclagen commented on services/CommentService.js:
```diff
@@ -0,0 +1,171 @@
+   * @param {number} [params.parentId]
+   * @returns {Promise<Comment>}
+   */
+  async create({ projectId, userId, message, parentId = null }) {
+    const projectExists = await this.Project.findByPk(projectId, { attributes: ["id"] });
+    if (!projectExists) {
+      const error = new Error("Project not found");
+      error.status = 404;
+      throw error;
+    }
+
+    if (parentId) {
+      const parentComment = await this.Comment.findByPk(parentId, { attributes: ["id", "projectId"] });
+      if (!parentComment) {
+        const error = new Error("Parent comment not found");
+        error.status = 404;
+        throw error;
+      }
+      if (parentComment.projectId !== projectId) {
+        const error = new Error("Parent comment does not belong to this project");
+        error.status = 400;
+        throw error;
+      }
+    }
+
+    return this.Comment.create({
+      projectId,
+      userId,
+      message,
+      parentId,
+    });
+  }
```
At the moment we have no means of viewing replies to replies so perhaps an additional check for the needs to be implement such as if the parentComment has a parentId then reject for the moment.

johnDavid97 replied:
This is just nitpicking, but I would change the name of the function to "createComment", since sequlize has an create method. In addtion, this is just how I would do it, but I would use regulaare params instead of an object as params, since you are creating an object inside of sequlizes creat method. Again just nitpicking but makes for a cleaner code. But this is great! Good standard backend service file, love it!

Anclagen replied:
This is the `CommentService`, so having a `create` method here is perfectly reasonable, when used in the controller as `commentService.create()`, its pretty self explanatory what is happening and their are no collisions to worry about.

As for destructuring vs regular params, I would say that’s subjective and context based. If you’ve only got a couple of params, regular ones are fine. But for something like a create function with multiple inputs, destructuring can be a valid approach as it avoids relying on param order and reduces the chance of silly bugs if things get reordered.

fmSand replied:
Made sure we only have one level reply for now. Good catch!
Changed the create name to differ from the create method. 
Left the destructing as is, ref what Alex says, and it being simpler to add more params if we want at a later point.

Anclagen commented on models/User.js:
```diff
@@ -100,5 +100,17 @@ module.exports = (sequelize, Sequelize) => {
       onDelete: "CASCADE",
     });
   };
+  // Anonymize comments when a user is soft-deleted:
+  User.addHook("afterDestroy", async (user, options) => {
+    await sequelize.models.Comment.update(
+      {
+        isDeleted: true,
+      },
+      {
+        where: { userId: user.id },
+        transaction: options.transaction,
+      }
+    );
+  });
```
This doesn't seem to be working, with our user delete function as it stands. Apparently this bypasses the hook
```js
  async softDelete(id) {
    return this.User.destroy({ where: { id } });
  }
```
The user delete would need to be set to something like this
```js
  async softDelete(id) {
    const user = await this.User.findByPk(id);
    return user.destroy();
  }
  ```
  
  Using a where in a destroy function on the user model is consider bulk and hence the hook doesn't trigger. We need to find by using findOne, or findByPk then use the destroy helper function attached to it to trigger the hook.

fmSand replied:
Will add this in it's own branch and pr

Anclagen commented on models/ProjectType.js:
This would of been better as a separate fix as this also break the Project outputs for their join types.

fmSand replied:
You're right. Changed it back

Anclagen reviewed (changes requested):
This is almost finished. Few things stood out to me:
- Breaking change on project-type: I think changing the model/table name should be done in its own PR so we can properly address all the potential issues it introduces.
- Comment serializer: I was a bit confused by the need for this function. The raw outputs already provide sufficient data for the frontend to handle formatting as needed.
- Hook doesn't trigger: The current soft delete implementation on the users endpoint doesn’t trigger the `afterDestroy` hook. This is because we're using `destroy({ where })`, which skips instance level hooks.
- Replying to replies: We currently allow this, but don't support the corresponding output format. We should prevent replying to replies for now until nesting is properly handled.

Great work on this everything else is working as intended. Once we iron out some problems we can merge this in with the dev branch.

johnDavid97 commented on controllers/commentController.js:
```diff
@@ -0,0 +1,77 @@
+var { getLimitOffset } = require("../utilities/getPagination");
+const serializeComment = require("../utilities/commentSerializer");
+
+async function createComment(req, res) {
+  const { message, parentId } = req.body;
+  const { projectId } = req.params;
+  const { id: userId } = req.user;
+
+  const comment = await commentService.create({ userId, projectId, message, parentId });
+
+  res.status(201).json(
+    successResponse({
+      message: "Comment created successfully",
+      data: comment,
+      statusCode: 201,
+    })
+  );
+}
```
This is great! Normally you would have simple methods inside of your Service file without errorhandlers and "if exists" and rather hav them in your controller this give a cleaner code, and takes away unenecsayr  errorhandlers.

fmSand replied:
Great feedback! Done

fmSand committed:
Merge branch 'dev' into feature/comments

fmSand committed:
refactor: address PR review and similar patterns to team code

fmSand committed:
fix: type error

fmSand committed:
move hook to own branch and PR

fmSand committed:
fix: add explicit status values to error responses in commentController

- Add 'fail' status to 404 error responses
- Add 'bad request' status to 400 error responses

fmSand committed:
refactor: standardize error handling in comment controller

- Replace direct error responses with createError utility
- Update error format to match new API standards
- Remove errorResponse import from controller
- Throw errors instead of returning responses for handling by global middleware

Anclagen reviewed (approved):
Looks good, nice work!

fmSand commented:
> Looks good, nice work!

@Anclagen Literally just committed changes to the error handling in the controller as you said this, if you don't mind checking over those real quick

fmSand committed:
add status

Anclagen commented:
I will check them once the errors refactor branch is merged as this pr is now dependent on #90 to function correctly

Anclagen reviewed (approved):
This all looks good now, I will merge it now.
